File: 0001-two-sum/0001-two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

 // One-pass Hash table, TC: O(n), SP: O(n)
var twoSum = function (nums, target) {

    const map = new Map();
    for (var i = 0; i < nums.length; i++) {
        const k = target - nums[i];
        if (map.has(k)) {
            return [i, map.get(k)];
        } else {
            map.set(nums[i], i);
        }
    }

    console.log("No two sum solution found");
    return [];
};


// Two-pass Hash table, TC: O(n), SP: O(n)
var twoSum2 = function (nums, target) {
    const n = nums.length; 
    const map = new Map();

    for(let i = 0; i < n; i++){
        map.set(nums[i], i);
    }

    for(let i = 0; i < n; i++){
        const k = target - nums[i];

        // the checking element should not be current element
        if(map.has(k) && map.get(k) !== i){ 
            return [i, map.get(k)];
        }
    }

    return [];
} 

// Brute force, TC: O(n^2), SP: O(1)
var twoSum1 = function (nums, target) {
    const n = nums.length; 

    for(let i = 0; i < n; i++){
    
        for(let j = i + 1; j < n; j++){
    
            if(nums[j] === target - nums[i]){
                return [i, j];
            }
        }
    }

    return [];
}

File: 0001-two-sum/0001-two-sum.py
class Solution:
    # HashMap Optimal solution
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hashMap = {}

        for i, num in enumerate(nums):
            compliment = target - num
            if compliment in hashMap:
                return [hashMap[compliment], i]
            hashMap[num] = i


    # Brute force, Time: O(N ^ 2)
    def twoSum_1(self, nums: List[int], target: int) -> List[int]:
        n = len(nums)

        for i in range(n):
            for j in range(i + 1, n):
                if nums[i] + nums[j] == target:
                    return [i, j]
                
        
        